### PR TITLE
Protocol specific ready checkers

### LINKED ISSUE #3941: What's the idea behind default SeldonDeployment readinessProbe for gRPC service
I get a feeling some container in my `SeldonDeployment` is not ready when upstream app is sending requests to it, seeing error like 

```
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.UNAVAILABLE
	details = "upstream connect error or disconnect/reset before headers. reset reason: connection failure"
	debug_error_string = "{"created":"@1645125206.023425744","description":"Error received from peer ipv4:10.48.13.4:5001","file":"external/com_github_grpc_grpc/src/core/lib/surface/call.cc","file_line":1063,"grpc_message":"upstream connect error or disconnect/reset before headers. reset reason: connection failure","grpc_status":14}"
```

but just for a few requests at the beginning.

And I see the default `readinessProbe` for each container in `SeldonDeployment` is like
```
    readinessProbe:
      failureThreshold: 3
      initialDelaySeconds: 20
      periodSeconds: 5
      successThreshold: 1
      tcpSocket:
        port: 9500
      timeoutSeconds: 1
```

I wonder why it's set this way, why use tcpSocket at port 9500 in particular? Why not use https://github.com/grpc-ecosystem/grpc-health-probe, which appears to be more commonly used in probing readiness of gRPC service?

### LINKED ISSUE #3945: Handle V2 ready checks in executor
* Keep tcp checks for Seldon and TF protocols

### PR DESCRIPTION
**Pull request by cliveseldon**
**What this PR does / why we need it**:

Add specific graph ready checkers to only be ready when underlying servers say they are ready rather than available

**Which issue(s) this PR fixes**:

Fixes #3945 
Fixes #3941 

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by seldondev** [comment]
@cliveseldon: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | 3edda197401a0a85d85332b605db8294f935cf64 | [link](http://dashboard-jx.34.121.152.94.nip.io/namespaces/jx/pipelineruns/seldon-core-pr-4028-integration-7tdkz) | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
@cliveseldon: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | 3edda197401a0a85d85332b605db8294f935cf64 | [link](http://dashboard-jx.34.121.152.94.nip.io/namespaces/jx/pipelineruns/seldon-core-pr-4028-integration-7tdkz) | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

File: executor/predictor/ready_checker.go
Comment (axsaucedo) [inline executor/predictor/ready_checker.go:16]: It seems that this URL won't work as it requires to be implemented (like the predict method) otherwise it errors out. A good solution would be to add a default behaviour to the funciton if the implementation is not found, that way we could also allow users to extend this which would be quite useful. The alternative (quick workaround) would be to use `/health/ping` instead, which would always return response, but it would not allow for users to extend it

File: executor/predictor/ready_checker.go
Comment (axsaucedo) [inline executor/predictor/ready_checker.go:18]: The tensoflow protocol does seem to have a way to query the model status/availability, but seems it's not fully standardised so could be something that can be explored further down the line

**Comment by seldondev** [comment]
@cliveseldon: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | 3edda197401a0a85d85332b605db8294f935cf64 | [link](http://dashboard-jx.34.121.152.94.nip.io/namespaces/jx/pipelineruns/seldon-core-pr-4028-integration-7tdkz) | `/test integration`
notebooks | 3edda197401a0a85d85332b605db8294f935cf64 | [link](http://dashboard-jx.34.121.152.94.nip.io/namespaces/jx/pipelineruns/io-seldon-core-pr-4028-notebooks-wggz9) | `/test notebooks`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
@cliveseldon: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | 3edda197401a0a85d85332b605db8294f935cf64 | [link](http://dashboard-jx.34.121.152.94.nip.io/namespaces/jx/pipelineruns/seldon-core-pr-4028-integration-7tdkz) | `/test integration`
notebooks | 3edda197401a0a85d85332b605db8294f935cf64 | [link](http://dashboard-jx.34.121.152.94.nip.io/namespaces/jx/pipelineruns/io-seldon-core-pr-4028-notebooks-wggz9) | `/test notebooks`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

File: executor/predictor/ready_checker.go
Comment (axsaucedo) [reply executor/predictor/ready_checker.go:16]: Actually it may be best to use `/health/ping` as this would ensure it has backwards compatibility, perhaps we could make the URL for the v1 server configurable via env var as well so that it's also possible to allow users to change the path if they do want to add custom extension

**Review by axsaucedo — CHANGES_REQUESTED** [review]
Added a couple of comments, it seems the URL being used for the v1 server needs further extensions in order for it to work by default

File: executor/predictor/ready_checker.go
Comment (cliveseldon) [reply executor/predictor/ready_checker.go:16]: It looks ok to me https://github.com/SeldonIO/seldon-core/blob/fd63953ea7e47048eebe5cff148c0053e90c4c7b/python/seldon_core/user_model.py#L512-L529

**Comment by axsaucedo** [comment]
Tested with the /health/ping and seems to work. 

On a (completely) separate note the test-rolling-update-ambas# failing tests are still failing, the errors are `504 gateway timeout` so thinking it may actually be on the ambassador side (also as no issues with istio) so thinking it may be unrelated to this readiness probe.

File: executor/predictor/ready_checker.go
Comment (axsaucedo) [reply executor/predictor/ready_checker.go:16]: Yeah you're right, I think it may just be that the tests are actually inheriting from the base class (which although mentioned in the docs it's not really common practice), which would basically throw an exception if not implemented

https://github.com/SeldonIO/seldon-core/blob/fd63953ea7e47048eebe5cff148c0053e90c4c7b/python/seldon_core/user_model.py#L140-L143

Comment (axsaucedo) [reply executor/predictor/ready_checker.go:16]: This seems to reflect the errors in the tests ```2022-04-04 19:19:37,408 - seldon_core.wrapper:handle_invalid_usage:60 - ERROR:  {'status': {'status': 1, 'info': 'health is not implemented', 'code': -1, 'reason': 'MICROSERVICE_BAD_DATA'}}```

File: helm-charts/seldon-core-operator/values.yaml
Comment (axsaucedo) [inline helm-charts/seldon-core-operator/values.yaml:71]: Perhaps it may be better to set to true by defualt with the updating pointing out to set it to false if required (ie they actually extend from the classes) - this may end up breaking some tests if they don't rebuild (although most of the test models should be rebuilt on the e2e test run). Or did you find that indeed quite a few tests were failing?

**Comment by axsaucedo** [comment]
Looks good - added a comment on whether best to have it as `true` by default, it also seems to have a conflict on the upgrading page, but other than lgtm

**Comment by axsaucedo** [comment]
All tests pass so can merge once conflict resolved (and open an issue to do the "deafult to true" point 

image

**Comment by axsaucedo** [comment]
@cliveseldon it seems there's still a merge conflict, it should be good to merge once that's resolved

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *[axsaucedo](https://github.com/SeldonIO/seldon-core/pull/4028#issuecomment-1104864259 "Approved")*

The full list of commands accepted by this bot can be found [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)~~ [axsaucedo]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**
